src/scripts/resource-channels.js

```diff
@@ -1,57 +1,70 @@
 const { mute } = require("./users/permissions");
 const { logMessageDate } = require("./utilities");
+const { User } = require("./users/user-utilities");
 
 const timeLimit = 120000;
 const warnOn = 3;
 const muteOn = 4;
+global.temp = {};
 
 // Spam Observer
-function resourcesObserver(message, counter, client) {
-	// if (message.member.hasPermission("MANAGE_ROLES")) return;
-	// Initializes user-specific variables if undefined.
-	counter[message.author.username + " " + message.author.id] = counter[message.author.username + " " + message.author.id] || {};
+function resourcesObserver(message, users, client) {
+	if (message.member.hasPermission("MANAGE_ROLES")) return;
 
-	counter[message.author.username + " " + message.author.id][message.channel.name] = counter[message.author.username + " " + message.author.id][message.channel.name] || {};
+	const id = message.author.id;
+	const name = message.author.username;
+	const channelName = message.channel.name;
 
-	counter[message.author.username + " " + message.author.id][message.channel.name].timeoutFlag = counter[message.author.username + " " + message.author.id][message.channel.name].timeoutFlag || false;
+	// Initializes new User instance if not defined.
+	users[id] = users[id] || new User(name, id);
+	users[id].addChannelMsg(message);
 
-	counter[message.author.username + " " + message.author.id][message.channel.name].count = counter[message.author.username + " " + message.author.id][message.channel.name].count || 0;
+	users[id].channels[channelName].count++;
 
-	let timeout;
-	function startTimeout() {
-		timeout = setTimeout(() => {
-			counter[message.author.username + " " + message.author.id][message.channel.name].count = 0;
-			counter[message.author.username + " " + message.author.id][message.channel.name].timeoutFlag = false;
-		}, timeLimit);
-	}
-	// Resets user-specific counter variable after timeout
-	if (counter[message.author.username + " " + message.author.id][message.channel.name].timeoutFlag === false) {
-		counter[message.author.username + " " + message.author.id][message.channel.name].timeoutFlag = true;
-		startTimeout();
-	} else {
-		clearTimeout(timeout);
-		startTimeout();
-	}
-
-	counter[message.author.username + " " + message.author.id][message.channel.name].count++;
-
-	console.log(`${message.author.username}'s Message number: ${counter[message.author.username + " " + message.author.id][message.channel.name].count}`);
+	const params = { users, id, channelName, message, client };
+	checkTimeoutFlag(params);
+	checkCount(params);
```

Comment:
I initially was passing them in separately, but with so many of the same parameters being passed for these functions I was worried about it being redundant and getting difficult to read. This is what the alternative looked like.
```javascript
checkTimeoutFlag(users, id, channelName);
checkCount(users, id, channelName, message, client);
	
function checkTimeoutFlag(users, id, channelName) {
	const userChannelKey = id + channelName; // Key to access appropriate timeout
	if (!users[id].channels[channelName].timeoutFlag) {
		users[id].channels[channelName].timeoutFlag = true;
		startTimeout(users, id, channelName, userChannelKey);
	} else {
		clearTimeout(timeouts[userChannelKey]);
		startTimeout(users, id, channelName, userChannelKey);
	}
}
	
function checkCount(users, id, channelName, message, client) {
	// ...
}
```
Here's what we have:
```javascript
const params = { users, id, channelName, message, client };
checkTimeoutFlag(params);
checkCount(params);

function checkTimeoutFlag(params) {
	const { users, id, channelName } = params;
	const userChannelKey = id + channelName; // Key to access appropriate timeout
	if (!users[id].channels[channelName].timeoutFlag) {
		users[id].channels[channelName].timeoutFlag = true;
		startTimeout(params, userChannelKey);
	} else {
		clearTimeout(timeouts[userChannelKey]);
		startTimeout(params, userChannelKey);
	}
}

function checkCount(params) {
	const { users, id, channelName, message, client } = params;
	// ...
}
```


Which do you think is cleaner?

Reply:
I think because of the packing and unpacking they both end up looking roughly the same, the only issue I have with it is that it makes it marginally more difficult to use/debug functions - anybody reading the function call has to find where the params are initialised to see what's actually being sent in and to see what it requires they would have to check inside the function rather than checking the function signature
In these cases it's just two functions very near to where they're called and the packing and unpacking happens right next to where they're needed so it's not unreasonable and I'm okay with it, but as a pattern I think it should be avoided

Reply:
Good points! I've made the changes you suggested so I'll go ahead and mark this as resolved.